### PR TITLE
Feat: Addon REST API

### PR DESCRIPTION
**Pull request by hongchaodeng**

### CODE REVIEW

**Review by Somefive — APPROVED** [review]

File: pkg/apiserver/rest/apis/v1/types.go [hunk 1]
```diff
@@ -20,6 +20,58 @@ import (
...
+	Icon string `json:"icon"`
+
+	Phase AddonPhase `json:"phase"`
+}
+
+// DetailAddonResponse defines the format for showing the addon details
+type DetailAddonResponse struct {
+	AddonMeta
+
+	Initializer string `json:"initializer"`
```
Comment (chivalryq) [inline pkg/apiserver/rest/apis/v1/types.go:72]: maybe `data`？This implement is going to swtiched.

**Review by chivalryq — CHANGES_REQUESTED** [review]

File: pkg/apiserver/rest/apis/v1/types.go [hunk 2]
```diff
@@ -20,6 +20,58 @@ import (
...
+
+// AddonStatusResponse defines the format of addon status response
+type AddonStatusResponse struct {
+}
+
+// AddonPhase defines the phase of an addon
+type AddonPhase string
+
+const (
+	// AddonPhaseDisabled indicates the addon is disabled
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:47]: What does disable mean here? Uninstall or logical disable?

File: pkg/apiserver/rest/apis/v1/types.go [hunk 3]
```diff
@@ -20,6 +20,58 @@ import (
...
+	Initializer string `json:"initializer"`
+}
+
+// ListAddonResponse defines the format for addon list response
+type ListAddonResponse struct {
+	Addons []AddonMeta `json:"addons"`
+}
+
+// AddonStatusResponse defines the format of addon status response
+type AddonStatusResponse struct {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:40]: Empty?

File: pkg/apiserver/rest/webservice/addon.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		Param(ws.PathParameter("name", "identifier of the addon").DataType("string")).
+		Writes(apis.AddonMeta{}))
+
+	ws.Route(ws.GET("/{name}").To(noop).
+		Doc("show details of an addon").
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Param(ws.PathParameter("name", "identifier of the addon").DataType("string")).
+		Writes(apis.DetailAddonResponse{}))
+
+	ws.Route(ws.GET("/{name}/status").To(noop).
```
Comment (barnettZQG) [inline pkg/apiserver/rest/webservice/addon.go:63]: This API doesn't make much sense, and its semantics can be directly translated into the previous API.

File: pkg/apiserver/rest/apis/v1/types.go [hunk 5]
```diff
@@ -20,6 +20,58 @@ import (
...
+
+	Icon string `json:"icon"`
+
+	Phase AddonPhase `json:"phase"`
+}
+
+// DetailAddonResponse defines the format for showing the addon details
+type DetailAddonResponse struct {
+	AddonMeta
+
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:71]: More addon information should be enriched, such as which XDefinition. Corresponding applications, etc.

File: pkg/apiserver/rest/apis/v1/catalog.go [hunk 6]
```diff
@@ -1,41 +0,0 @@
...
-*/
-
-package v1
-
-import (
-	"github.com/oam-dev/kubevela/pkg/apiserver/model"
-)
-
-// CatalogRequest defines the body of catalog request
-type CatalogRequest struct {
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply pkg/apiserver/rest/apis/v1/catalog.go:24]: Not needed.

File: pkg/apiserver/rest/apis/v1/types.go (hunk 2 above)
Comment (hongchaodeng) [reply pkg/apiserver/rest/apis/v1/types.go:47]: uninstall. Following the CLI

File: pkg/apiserver/rest/apis/v1/types.go (hunk 3 above)
Comment (hongchaodeng) [reply pkg/apiserver/rest/apis/v1/types.go:40]: Going to add more.

File: pkg/apiserver/rest/webservice/addon.go (hunk 4 above)
Comment (hongchaodeng) [reply pkg/apiserver/rest/webservice/addon.go:63]: They are not equivalent.
GET /:name returns the full data of the Addon. This is for showing the full page.
GET /:name/status returns the status only. This is for polling the installation status.

File: pkg/apiserver/rest/apis/v1/types.go (hunk 5 above)
Comment (hongchaodeng) [reply pkg/apiserver/rest/apis/v1/types.go:71]: Adding.

File: pkg/apiserver/rest/apis/v1/types.go [hunk 7]
```diff
@@ -20,6 +20,58 @@ import (
...
 
+// CreateAddonRequest defines the format for addon create request
+type CreateAddonRequest struct {
+	Name string `json:"name" validate:"required"`
+
+	Description string `json:"description,omitempty"`
+
+	Icon string `json:"icon"`
+
+	Initializer string `json:"initializer"`
```
[the remark this answers is not in the document]
Comment (hongchaodeng) [reply pkg/apiserver/rest/apis/v1/types.go:31]: Load the Initializer object to deploy

**Review by barnettZQG — APPROVED** [review]
LGTM

**Review by wonderflow — APPROVED** [review]

**Pull request merged**